kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/partition/DynamicPartitioner.java

```diff
@@ -0,0 +1,256 @@
+   * the batch size.
+   */
+  private void establishedWriteAll(TableId table,
+                                   List<InsertAllRequest.RowToInsert> elements,
+                                   String topic,
+                                   Set<Schema> schemas) throws BigQueryConnectException,
+                                                               InterruptedException {
+    int currentIndex = 0;
+    while (currentIndex < elements.size()) {
+      try {
+        // very similar to equalPartitioner
+        // handle case where no partitioning is necessary:
+        if (elements.size() <= currentBatchSize) {
+          writer.writeRows(table, elements, topic, schemas);
+          currentIndex = elements.size();
+          return;
+        }
+
+        int numPartitions = (int)Math.ceil(elements.size() / (currentBatchSize * 1.0));
+        int minBatchSize = elements.size() / numPartitions;
+        int spareRows = elements.size() % numPartitions;
+
+        for (int partition = 0; partition < numPartitions; partition++) {
+          // the first spareRows partitions have an extra row in them.
+          int batchSize = partition < spareRows ? minBatchSize + 1 : minBatchSize;
+          int endIndex = currentIndex + batchSize;
+          writer.writeRows(table, elements.subList(currentIndex, endIndex), topic, schemas);
+          currentIndex = endIndex;
+        }
+      } catch (BigQueryException exception) {
+        if (isPartitioningError(exception)) {
+          // immediately decrease batch size and try again with remaining elements.
+          decreaseBatchSize();
+          contSuccessCount = 0;
+          logger.debug("Partition error during establishedWriteAll, reducing batch size to {}",
+                       currentBatchSize);
+          continue;
+        } else {
+          throw new BigQueryConnectException(
+              String.format("Failed to write to BigQuery table %s", table),
+              exception);
+        }
+      }
+    }
+
+    contSuccessCount++;
+    if (contSuccessCount >= CONT_SUCCESS_COUNT_BUMP) {
+      logger.debug("{} successful establishedWriteAlls in a row, increasing batchSize to {}",
+                   contSuccessCount, currentBatchSize);
+      increaseBatchSize();
+    }
+  }
+
+  private boolean isPartitioningError(BigQueryException exception) {
+    if (exception.code() == BAD_REQUEST_CODE
+        && exception.error() == null
+        && exception.reason() == null) {
+      // more than 10MB/req
+      // todo google may add an error/reason at some point.
+      // They probably won't, but it would be nice.
+      return true;
+    }
+    return false;
+  }
+
+  // for some reason when I run tests it's asking for these to be public?? I'm not sure why...
+  private void increaseBatchSize() {
+    currentBatchSize = Math.min(currentBatchSize * 2, MAXIMUM_BATCH_SIZE);
+  }
+
+  private void decreaseBatchSize() {
+    currentBatchSize = currentBatchSize / 2;
```

Comment:
Or maybe we should puke if we're getting msg too large errors from Google with a batch size of 1?